File: src/main/java/com/github/joelws/release/tracker/common/GenericDaoImpl.java
package com.github.joelws.release.tracker.common;

import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

@Repository public class GenericDaoImpl<T, PK> implements GenericDao<T, PK>
{
    @PersistenceContext private EntityManager entityManager;

    private Class<T> tClass;

    public GenericDaoImpl(Class<T> tClass)
    {
        this.tClass = tClass;
    }

    @Transactional @Override public T create(T type)
    {
        entityManager.persist(type);
        return type;
    }

    public GenericDaoImpl()
    {
    }

    @Transactional @Override public T read(PK primaryKey)
    {
        return entityManager.find(tClass, primaryKey);
    }

    @Transactional @Override public T update(T type)
    {
        entityManager.merge(type);
        return type;
    }

    @Transactional @Override public void delete(PK primaryKey)
    {
        entityManager.remove(entityManager.getReference(tClass, primaryKey));
    }

    protected EntityManager getEntityManager()
    {
        return entityManager;
    }

    protected void setEntityManager(EntityManager entityManager)
    {
        this.entityManager = entityManager;
    }
}

File: src/main/kotlin/com/joelws/release/tracker/dao/GenericDaoImpl.kt
package com.joelws.release.tracker.dao

import org.funktionale.option.Option
import org.funktionale.option.toOption
import java.lang.reflect.ParameterizedType
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext


abstract class GenericDaoImpl<T : Any, PK> : GenericDao<T, PK> {

    @PersistenceContext
    lateinit var entityManager: EntityManager

    private val klass: Class<T>

    init {
        val pt = javaClass.genericSuperclass as ParameterizedType
        @Suppress("UNCHECKED_CAST")
        val klass = pt.actualTypeArguments[0] as Class<T>
        this.klass = klass
    }

    override fun create(entity: T): Option<T> {
        entityManager.persist(entity)
        return entity.toOption()
    }

    override fun read(identifier: PK): Option<T> {
        return entityManager.find(klass, identifier).toOption()
    }

    override fun update(entity: T): Option<T> {
        return entityManager.merge(entity).toOption()
    }

    override fun delete(identifier: PK) {
        entityManager.remove(entityManager.getReference(klass, identifier))
    }

}
